locator/tablets.cc

```diff
@@ -245,9 +246,26 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+tablet_replica tablet_map::get_primary_replica(tablet_id id) const {
     const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+    auto replicas = boost::copy_range<std::vector<tablet_replica>>(info.replicas);
```

Comment:
How about sorting when replicas is updated? We might use get_primary_replica in data path later.

Reply:
We can't do that, unless MV has a better way to pair replicas in the base table with the respective ones in the view.
See https://github.com/scylladb/scylladb/issues/17147, https://github.com/scylladb/scylladb/issues/17043, https://github.com/scylladb/scylladb/issues/17265

Reply:
> How about sorting when replicas is updated? We might use get_primary_replica in data path later.

Regarding the primary replica in the data path.
We don't have a crisp design for that yet, but my assumption is that when we have alignment between raft groups and tablets, e.g. for LWT over raft, we will have a raft group leader per tablet and it can be seen as the primary replica.  With hierarchical raft, you'd also have a sub-group leader per dc, and that could be seen as the primary replica within the dc.

Cc @tgrabiec

Reply:
I see.

Reply:
Why the copy? And if the copy is needed, why not just:
```c++
auto replicas = std::vector(info.replicas.begin(), info.replicas.end());
```
Let's keep things simple.

Reply:
As I answered [above](https://github.com/scylladb/scylladb/pull/18784/files#r1609386314), the copy is needed since we mustn't sort the replicas in place.

The constructor you suggested makes sense and it is indeed simple.
I'll send a new version, also with the partial sort that should be more efficient on average for this purpose, and with the expected relatively low number of replicas.
replica count is number of dcs * rf_per_dc and I think we can safely assume this will be bound by 5 * 5 = 25 replicas, in very large installations